packages/flutter/lib/src/gestures/multitap.dart

```diff
@@ -587,3 +587,403 @@ class MultiTapGestureRecognizer extends GestureRecognizer {
+/// `DoubleTapGestureRecognizer`, and the user double-taps.
+///
+/// For example, consider a list item that should be _selected_ on the first
+/// tap and _cause an edit dialog to open_ on a double-tap. If you use both
+/// [TapGestureRecognizer] and [DoubleTapGestureRecognizer], there are a few
+/// problems: (1) if the user single-taps the list item, it will not select
+/// the list item until after enough time has passed to rule out a double-tap,
+/// and (2) if the user double-taps the list item, it will not select the list
+/// item at all. The solution is to use [SerialTapGestureRecognizer] and use
+/// the tap count to either select the list item or open the edit dialog.
+class SerialTapGestureRecognizer extends GestureRecognizer {
+  /// Creates a serial tap gesture recognizer.
+  SerialTapGestureRecognizer({
+    Object? debugOwner,
+    Set<PointerDeviceKind>? supportedDevices,
+  }) : super(debugOwner: debugOwner, supportedDevices: supportedDevices);
+
+  /// A pointer has contacted the screen at a particular location\, which might
+  /// be the start of a serial tap.
+  ///
+  /// If this recognizer loses the arena before the serial tap is completed
+  /// (either because the gesture does not end up being a tap or because another
+  /// recognizer such a [TapGestureRecognizer] or [DoubleTapGestureRecognizer]
+  /// wins outright), [onSerialTapCancel] is called next. Otherwise,
+  /// [onSerialTapUp] is called next.
+  ///
+  /// The [SerialTapDownDetails.count] that is passed to this callback
+  /// specifies the series tap count.
+  GestureSerialTapDownCallback? onSerialTapDown;
+
+  /// A pointer that previously triggered [onSerialTapDown] will not end up
+  /// triggering the corresponding [onSerialTapUp].
+  ///
+  /// If the user completes the serial tap, [onSerialTapUp] is called instead.
+  ///
+  /// The [SerialTapCancelDetails.count] that is passed to this callback will
+  /// match the [SerialTapDownDetails.count] that was passed to the
+  /// [onSerialTapDown] callback.
+  GestureSerialTapCancelCallback? onSerialTapCancel;
+
+  /// A pointer has stopped contacting the screen at a particular location,
+  /// representing a serial tap.
+  ///
+  /// If the user didn't complete the tap, or if another recognizer won the
+  /// arena before this recognizer handled the completed tap, then
+  /// [onSerialTapCancel] is called instead.
+  ///
+  /// The [SerialTapUpDetails.count] that is passed to this callback
+  /// specifies the series tap count.
+  ///
+  /// Because this callback fires for every completed serial tap, this callback
+  /// may fire even when this recognizer doesn't win the arena. For example, if
+  /// if it is competing with [DoubleTapGestureRecognizer], both recognizers
+  /// will remain in the arena after the first tap, and [onSerialTapUp] will
+  /// fire with a [SerialTapUpDetails.count] of 1. Then, after the second tap,
+  /// the double-tap recognizer will win the arena.
```

Comment:
I believe whether tapUp or cancel fires now depends on the order the `SerialTapGestureRecognizer` and the `DoubleTapGestureRecognizer` entered the arena. If the DTGR enters first, it will see the up-2 event first, declare victory and cause the STGR to call `cancel`. If the STGR entered the arena first, it gets to see the up-2 first, calls tapUp, then the DTGR wins.

I find the semantics of when tapUp and when cancel gets called a little confusing. If the STGR competes with a regular TapRecognizer, the STGR will always call down/up (even when the TR wins). However, if it competes with a DTGR it may under certain circumstances call down/cancel for the second tap.

Reply:
This is fixed now - and I added test cases to match.